cmd/node-agent/main.go

```diff
@@ -0,0 +1,220 @@
+				return client.Core().Nodes().Watch(opts)
+			},
+		},
+		&v1.Node{},
+		30*time.Second,
+		framework.ResourceEventHandlerFuncs{
+			UpdateFunc: nodeUpdateCallback,
+		},
+	)
+
+	var err error
+	sysdConn, err = dbus.New()
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	for {
+		// We say to never stop, but this method will
+		// return if it fails to contact an API server.
+		// This service should be resilient to temporary
+		// loss of an API server.
+		nodeController.Run(wait.NeverStop)
+		time.Sleep(time.Second)
+	}
+}
+
+func newAPIClient() clientset.Interface {
+	config, err := restclient.InClusterConfig()
+	if err != nil {
+		glog.Fatal(err)
+	}
+	return clientset.NewForConfigOrDie(config)
+}
+
+func nodeUpdateCallback(oldObj, newObj interface{}) {
+	newNode := newObj.(*v1.Node)
+	ans := newNode.Annotations
+	glog.Infof("begin node update callback. Annotations: %#v", ans)
+	if ans[desiredVersionAnnotation] != ans[currentVersionAnnotation] {
+		glog.Infof("node version update: %s", ans[desiredVersionAnnotation])
+		updateNodeVersion(ans[desiredVersionAnnotation], newNode)
+	}
+	if ans[desiredKubeletConfigAnnotation] != ans[currentKubeletConfigAnnotation] {
+		glog.Infof("node config update: %s", ans[desiredKubeletConfigAnnotation])
+		updateKubeletConfig(ans[desiredKubeletConfigAnnotation], newNode)
+	}
+}
+
+func updateNodeVersion(version string, node *v1.Node) {
+	uo := unit.NewUnitOption(
+		"Service",
+		"Environment",
+		fmt.Sprintf("KUBELET_VERSION=%s", version),
+	)
+	verifyAPIServerVersionGreaterOrEqual(version)
+	replaceSystemdDropInAndRestartUnit(kubeletVersionServicePath, kubeletVersionService, []*unit.UnitOption{uo})
+	node.Annotations[currentVersionAnnotation] = version
+	updateNode(node)
+}
+
+func verifyAPIServerVersionGreaterOrEqual(version string) {
+	glog.Info("verifying that kubelet version <= API server version")
+
+	vi, err := client.Discovery().ServerVersion()
+	if err != nil {
+		glog.Fatalf("Unable to get server version: %v", err)
+	}
+	apiVer, err := semver.Parse(vi.GitVersion)
+	if err != nil {
+		glog.Fatalf("Unable to get server version: %v", err)
+	}
+	kubeletVer, err := semver.Parse(strings.Replace(version, "_", "+", 1))
+	if err != nil {
+		glog.Fatalf("Unable to get server version: %v", err)
+	}
+	if kubeletVer.GT(apiVer) {
+		glog.Fatalf("Kubelet version %s is greater than API Server version %s", version, vi.GitVersion)
+	}
+}
+
+func updateKubeletConfig(configMapName string, node *v1.Node) {
+	flags := configMapToCommandLineFlags(configMapName)
+	uo := unit.NewUnitOption(
+		"Service",
+		"Environment",
+		fmt.Sprintf("'KUBELET_FLAGS=%s'", flags),
+	)
+	replaceSystemdDropInAndRestartUnit(kubeletConfigServicePath, kubeletConfigService, []*unit.UnitOption{uo})
+	node.Annotations[currentKubeletConfigAnnotation] = configMapName
+	updateNode(node)
+}
+
+func updateNode(node *v1.Node) {
+	_, err := client.Core().Nodes().Update(node)
+	if err != nil {
+		glog.Fatal(err)
```

Comment:
I initially di that to make it obvious when something is wrong (you can see clearly with `get pods` or something that it's failing). I can drop this down a bit though.